design/include-cleaner.md

```diff
@@ -0,0 +1,187 @@
+provide diagnostics to keep includes in an
+[IWYU](https://include-what-you-use.org/)-clean state.
+
+Include Cleaner is available in "preview" mode with an incomplete set of
+capabilities and can be enabled through [configuration
+file](/config#UnusedIncludes). If you experience any bugs, please submit a bug
+report in [clangd/issues](https://github.com/clangd/clangd/issues).
+
+{:.v14}
+
+## How it works
+
+Include Cleaner issues diagnostics for includes that are present but not used
+in the main file. When you open a file, clangd will analyze the symbols the
+file uses and mark all headers defining these symbols as "used". The warnings
+will be issued for the headers that are included but not marked as "used".
+Example:
+
+```c++
+// foo.h
+class Foo {};
+```
+
+```c++
+// bar.h
+class Bar {};
+```
+
+```c++
+// main.cpp
+
+#include "foo.h"
+#include "bar.h" // <- Will be marked as unused and suggested to be removed.
+
+int main() {
+  Foo f;
+}
+```
+
+Here, `main.cpp` only makes use of symbols from `foo.h` and removing `#include
+"bar.h"` prevents unnecessary parsing of `bar.h` and allows breaking the
+dependency on it.
+
+### Deciding what headers are used
+
+Clangd relies on Clang AST to deduce which headers are used and which aren't,
+the whole Include Cleaner decision process is described below.
+
+#### Scanning the main file
+
+IncludeCleaner will traverse Clang AST of the main file (the file currently
+opened in the editor). It will recursively visit AST nodes and collect locations
+of all referenced symbols (e.g.  types, functions, global variables). Any
+declaration explicitly mentioned in the code, brought in via macro expansions,
+implicitly through type deductions or template instantiations will be marked as
+"used". Example:
+
+```c++
+// foo.h
+
+// USED
+int foo();
+
+// USED
+#define FOO foo
+
+// USED
+struct Bar {
+  Bar();
+}
+
+// USED
+struct Baz;
+
+// USED
+template <typename T> Baz getBaz();
+```
+
+```c++
+// main.cpp
+
+#include "foo.h"
+
+int main() {
+  // Uses foo() and FOO
+  FOO();
+  // Uses Baz, getBaz and Bar.
+  auto baz = getBaz<Bar>();
+}
+```
+
+This means that Include Cleaner is conservatively treating symbols in the
+expanded code as usages as opposed to only explicitly spelled symbols.
+
+Include Cleaner will also traverse the macro names in the spelled code to
+collect used macros.
+
+After the process is complete, all declaration and definition locations will be
+collected and passed to the next stage.
+
+#### Marking the headers as used
+
+`SourceLocation` instances collected at the previous step will be converted to
+`FileID`s and deduplicated. In this stage, it is important to attribute the
+locations in some headers to their includes. Some of the `FileID`s correspond to
+non self-contained headers, meaning the user should actually include their
+parent rather than the header itself. Other important headers with this property
```

Comment:
this might not necessarily be `parent`, maybe `ancestor` but I think mentioning `the first self-contained header in the include stack` makes it a lot more clear (despite being worthy)

Reply:
oh now that I read the rest, i suppose this is just an artifact of the past. maybe drop everything but the first sentence of this paragraph ?